File: Data Structure/tree/tree.py
class TreeNode:
  def __init__(self, val=0):
    self.val = val
    self.left = None
    self.right = None

node1 = TreeNode(1)
node2 = TreeNode(2)
node3 = TreeNode(3)
node4 = TreeNode(4)
node5 = TreeNode(5)
node6 = TreeNode(6)
node7 = TreeNode(7)

node1.left = node2
node1.right = node3
node2.left = node4
node2.right = node5
node3.left = node6
node3.right = node7

def recursivePreorder(node): 
  if node is None:
    return 
  print(node.val, end=' ')
  recursivePreorder(node.left) 
  recursivePreorder(node.right) 

recursivePreorder(node1)

def recursiveInorder(node): 
  if node is None:
    return 
  recursiveInorder(node.left) 
  print(node.val, end=' ')
  recursiveInorder(node.right) 

recursiveInorder(node1)

def recursivePostorder(node): 
  if node is None:
    return 
  recursivePostorder(node.left) 
  recursivePostorder(node.right)
  print(node.val, end=' ')

recursivePostorder(node1)

File: Data Structure/tree/tree.js
// 트리 정의하기
class Tree {
  constructor(val) {
    this.val = val;
    this.leftNode = null;
    this.rightNode = null;
  }

  setVal(val) {
    this.val = val;
  }

  setLeft(node) {
    this.leftNode = node;
  }

  setRight(node) {
    this.rightNode = node;
  }

  // in-order
  inOrderTree(node) {
    if (!node) {
      return;
    }
    this.inOrderTree(node.leftNode);
    console.log(node.val);
    this.inOrderTree(node.rightNode);
  }

  // pre-order
  preOrderTree(node) {
    if (!node) {
      return;
    }

    console.log(node.val);
    this.preOrderTree(node.leftNode);
    this.preOrderTree(node.rightNode);
  }

  // post-order
  postOrderTree(node) {
    if (!node) {
      return;
    }

    this.postOrderTree(node.leftNode);
    this.postOrderTree(node.rightNode);
    console.log(node.val);
  }
}

// 부모, 자식 관계 정의
let root = new Tree(1);
let node2 = new Tree(2);
let node3 = new Tree(3);
let node4 = new Tree(4);
let node5 = new Tree(5);
let node6 = new Tree(6);
let node7 = new Tree(7);
root.setLeft(node2);
root.setRight(node3);
root.leftNode.setLeft(node4);
root.leftNode.setRight(node5);
root.rightNode.setLeft(node6);
root.rightNode.setRight(node7);

// 출력하기
console.log('>>>> InOrder Start!! ');
root.inOrderTree(root);

console.log('>>>> preOrder Start!! ');
root.preOrderTree(root);

console.log('>>>> postOrder Start!! ');
root.postOrderTree(root);

var preorderTraversal = function (root) {
  /**
   * Algorithm:
   * 1. Create an empty stack [];
   * 2. Do while stack is not empty:
   * 2.1. Pop an item from stack and add it to the 'result' array.
   * 2.2. Push 'right child' of popped item to stack.
   * 2.3. Push 'left child' of popped item to stack.
   */
  if (root == null) {
    return [];
  }

  const stack = [];
  const result = [];

  stack.push(root);

  while (stack.length > 0) {
    let current = stack.pop();
    result.push(current.val);

    if (current.right) stack.push(current.right);
    if (current.left) stack.push(current.left);
  }

  return result;
};